src/masternodes/mn_checks.cpp

```diff
@@ -3844,6 +3844,30 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor {
             if (!res)
                 return res;
         } else if (obj.type == CTransferBalanceType::EvmOut) {
+            for (const auto& [addr, _] : obj.from) {
+                CTxDestination dest;
+                if (ExtractDestination(addr, dest)) {
+                    if (dest.index() != WitV16KeyEthHashType) {
+                        return Res::Err("From address must be an ETH address in case of \"evmout\" transfertype");
+                    }
+                }
+                bool foundAuth = false;
+                for (const auto &input : tx.vin) {
+                    const Coin &coin = coins.AccessCoin(input.prevout);
+                    std::vector<TBytes> vRet;
+                    if (Solver(coin.out.scriptPubKey, vRet) == txnouttype::TX_PUBKEYHASH)
+                    {
+                        auto it = input.scriptSig.begin();
+                        CPubKey pubkey(input.scriptSig.begin() + *it + 2, input.scriptSig.end());
+                        auto script = GetScriptForDestination(WitnessV16EthHash(pubkey));
+                        if (script == addr)
+                            foundAuth = true;
+                    }
+                }
+                if (!foundAuth)
+                    return Res::Err("authorization not found for %s in the tx", ScriptToString(addr));
+            }
```

Comment:
Why not do this at line 3813 where the Auth for the other cases is checked?

Reply:
Why would you loop two times? you check Auth and Eth-type in the "Auth section" so you do not have to check here.

Would be clearer code-wise:
* check valid auth and address type
* check values and sums
* if all good: do actual transfer.

Reply:
It doesn't matter where it is done. I wanted to do it in the part for EVMOut if section, but because other auths are the same I didn't do them in their own section but in the beginning.